### PR TITLE
CRM-20264: additional fixes for online Event registration, backoffice, membership and other validation

### PR DESCRIPTION
**Pull request by monishdeb**
* [CRM-20264: Store CC type and last 4 digits from Contribution form](https://issues.civicrm.org/jira/browse/CRM-20264)

### CODE REVIEW

**Comment by monishdeb** [comment]
@eileenmcnaughton @colemanw @JoeMurray I have fixed some remaining issues reported by @pradpnayak after testing against rc. And these are the issues fixed in this PR:
1. **CC type ID and number isn't recorded after doing online event registration** (using dummy processor). This is the [fix](https://github.com/civicrm/civicrm-core/pull/10234/files#diff-7a2ee8f032ba2340f94769e7a164e587R964) in accordance with https://github.com/civicrm/civicrm-core/pull/10157/files#diff-ecead08fd9d62dd134a679e60a2ba6b4R2040 
2.  **CC type ID and number isn't recorded after doing backoffice membership payment in live mode.** 
[Fixed](https://github.com/civicrm/civicrm-core/pull/10234/files#diff-fa0276456bc05f4629f775e6846f62edR1360) and also added unit test
3. **Invalid pan_truncation number can be entered by allowing less than 4 digits last CC number**. Fixed [here](https://github.com/civicrm/civicrm-core/pull/10234/files#diff-f8705625754a6a2f3655a8c7c696969cR794)

File: CRM/Member/Form/Membership.php [hunk 1]
```diff
@@ -1356,6 +1356,10 @@ public function submit() {
     if ($this->_mode) {
       $params['total_amount'] = CRM_Utils_Array::value('total_amount', $formValues, 0);
 
+      //CRM-20264 : Store CC type and number (last 4 digit) during backoffice or online payment
```
Comment (eileenmcnaughton) [inline CRM/Member/Form/Membership.php]: the fact this is inside $this->_mode gives me pause, also I feel we can use a parent function. The changes to Confirm & to Core_Payment look correct.

Comment (eileenmcnaughton) [inline CRM/Member/Form/Membership.php]: Shouldn't     $this->beginPostProcess(); catch this ?

Comment (monishdeb) [inline CRM/Member/Form/Membership.php]: Issue is at here https://github.com/civicrm/civicrm-core/pull/10234/files#diff-fa0276456bc05f4629f775e6846f62edR1316 where ```$params``` is taking specific attribute if ```record_contribution = TRUE``` which is ONLY available at backoffice form not in live mode. So I found as this fix is the win-win solution.

**Comment by eileenmcnaughton** [comment]
OK - yes I agree that's OK on the member form. I kind of feel like pushing a bit further & taking from $this->_params rather than $formValues further down, but I think this is ok - merge on pass

**Pull request merged**